### PR TITLE
Issue#1643: Coinselection prunes extraneous inputs from ApproximateBestSubset

### PR DESCRIPTION
**Pull request by murchandamus**
Improvement for [Issue#1643](https://github.com/bitcoin/bitcoin/issues/1643):
A further pass over the available inputs has been added to ApproximateBestSubset after a candidate set has been found. It will prune any extraneous inputs in the selected subset, in order to decrease the number of inputs and the resulting change.

### CODE REVIEW

**Comment by RHavar** [comment]
This dataset might be useful for any simulation: https://gist.github.com/RHavar/7cd6f3fcf2bd3e485458

This is real world data from MoneyPot.com that shows receive (positive amounts) and send (negative amount) over time (oldest at the top, newest at the bottom) denominated in bitcoins.

**Comment by jgarzik** [comment]
This PR has been open for a while, but garnered no ACKs.  The author seems to have put a fair amount of time and thought into it.  However, this definitely needs more review and testing.

Ping, maintainers/testers/helpers?

**Comment by Xekyo** [comment]
@jgarzik: Hi Jeff,
Thanks for the ping. 
I had lost track of this, but I see now that @RHavar posted data that might help to explore what the patch would do with some realistic data for transaction sizes. I'll have a look this weekend.

**Comment by Xekyo** [comment]
I've created another testcase using the MoneyPot.com data. On that one, I get different results yet again, this time the Regular wallet has the least UTXO in the end.

Looking over the results, I noticed that all wallets created a change output in the single digit satoshis. Do I remember correctly that the wallet shouldn't create Dust Outputs? Shouldn't the change be at least 540 satoshis? If so, I should probably fix that behavior still and then have another look.

However, if it ends up looking as promising as today, I would propose to just expire this Pull-Request.

**Comment by MarcoFalke** [comment]
@Xekyo:
> prune extraneous inputs

Wouldn't "extraneous inputs" imply an issue in the coin selection algorithm which should be fixed instead? I invite you to look at #6696 which is a different approach. I am happy to shoot up some simulations if anyone is interested. (Ping me at the other PR)

**Comment by laanwj** [comment]
From what I understand ApproximateBestSubset is an approximate algorithm for the following:
```
Input: [sizes], nLow, nTarget
Output: X=subset([sizes]) for which nTargetValue <= sum(X) < nLow
        and sum(X)-nTargetValue as small as possible
```
The "approximate" part means that it may select a solution which is not the optimal one (e.g. sum(X)-nTargetValue is not really as small as possible).

Your fix drops elements i from the result X for which sum(X)-i is still >= nTargetvalue, which if possible leads to a trivially better solution.

From what I understand this can improve the result because vValue is sorted from low to high, and elements are added in that order, it can happen that an element is added to get the sum above nTargetValue, but makes earlier-added small elements redundant.

- What about moving the post-processing step to the end? This removes the performance impact per checked subset, and still makes sure redundant outputs are removed from the final result.

I've managed to reproduce this as well. For e.g. these input values:
```
vValues = [10, 10, 10, 10, 10, 10, 10, 10, 10, 100, 1000]
nTargetValue = 251
```
ApproximateBestSubset sometimes selects `[10, 10, 1000]`. Postprocessing like this can drop the redundant `10` inputs.

Concept ACK - this cannot make things worse. Would be nice to have a unit test.

**Comment by MarcoFalke** [comment]
@laanwj Great that you had a look at this.

> vValue is sorted from low to high ...

Guess I missed that during review.

> unit test

@Xekyo are you still working on this? Should I take over?

**Comment by Xekyo** [comment]
Hi,

Am 20.11.2015 11:11 schrieb "MarcoFalke" <notifications@github.com>:
> @Xekyo are you still working on this? Should I take over?

I'd be interested to take another look, but I'm currently traveling. Please
feel free to take over. Otherwise, I'll check it out in December.

Xekyo

**Comment by Xekyo** [comment]
@laanwj I've edited my fork to move the post-processing step to the end of ApproximateBestSubset. However, this patch may cause fewer dust outputs to be spent which contradicts [Greg's assessment](https://github.com/bitcoin/bitcoin/pull/4906#issuecomment-59698734) above. Are you sure about "cannot make things work"? I feel my simulations have been somewhat inconclusive on that point.

**Comment by Xekyo** [comment]
@MarcoFalke I realized my mistake and fixed it. After all my commits, I did the rebase as you suggested, and pushed to "Fix-#1643". I hope I did that right. :)

**Comment by Xekyo** [comment]
I've added a simple test with

       vValues = [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 100, 100]
       nTargetValue = 222

with an expected `nBest == 230` on one iteration.

Sorry, I'm not sure how I would run the test on my machine right now, so I figured I'd just push it. I'll check back tomorrow.

File: src/test/coinselection_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#include "test/test_bitcoin.h"
+#include "wallet.cpp"
+
+#include <boost/test/unit_test.hpp>
+
+using namespace std;
+
+BOOST_FIXTURE_TEST_SUITE(coinselection_tests, BasicTestingSetup)
+
+BOOST_AUTO_TEST_CASE(sanity)
```
Comment (MarcoFalke) [inline src/test/coinselection_tests.cpp]: I think you can just add this to `src/wallet/test/wallet_tests.cpp`.

Comment (Xekyo) [inline src/test/coinselection_tests.cpp]: Ah thanks, I had been wondering why there was no wallet tests in the regular test directory.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -1625,13 +1625,21 @@ static void ApproximateBestSubset(vector<pair<CAmount, pair<const CWalletTx*,uns
...
                     }
                 }
             }
         }
     }
+
+    //Reduces the approximate best subset by removing any inputs that are smaller than the surplus of nTotal beyond nTargetValue. 
+    for (unsigned int i = 0; i < vValue.size(); i++)
+    {                        
+	if (vfBest[i] && (nBest - vValue[i].first) >= nTargetValue )
```
Comment (laanwj) [inline src/wallet/wallet.cpp]: Nit: indentation

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -1625,13 +1625,21 @@ static void ApproximateBestSubset(vector<pair<CAmount, pair<const CWalletTx*,uns
                             nBest = nTotal;
                             vfBest = vfIncluded;
                         }
-                        nTotal -= vValue[i].first;
```
Comment (laanwj) [inline src/wallet/wallet.cpp]: This (along with `&& !fReachedTarget` is a new, different change to the algorithm than the post-processing step initially proposed.

I think we should separate these, the postprocessing to remove extraneous outputs below is obviously correct, but I'm not sure about this.

Comment (Xekyo) [inline src/wallet/wallet.cpp]: Fair enough, I'll take it out and create another PR later.

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (Xekyo) [inline src/wallet/wallet.cpp]: I've replaced the tab with four spaces.

**Comment by laanwj** [comment]
ACK

**Comment by Xekyo** [comment]
I rebased and used fixup on the commits that only fixed previous mistakes.

**Comment by Xekyo** [comment]
@laanwj Travis had cleared my patch just when I realized that I had also messed up the indentation on the test. I've rebased it into two a commit for the code, and one for the test. I expect this to turn green again. Sorry for keeping travis so busy. :(

Is there anything else that needs to be done about this PR?

**Comment by juggalo71** [comment]
Penis....

**Comment by MarcoFalke** [comment]
Wallet code itself looks good! Post-merge Tested ACK.

@Xekyo Thanks for sticking with this so long! There seems to be a small issue with the test code but I will create another PR for this...

Note @laanwj: vValue is **not** sorted from low to high but from high to low but I think you meant it the right way. ;)

**Comment by MarcoFalke** [comment]
Oh, and you mentioned this will improve cases such as #1643. Which is not true, at least for this very transaction mentioned in #1643.

Still, I assume pruning will help for really large wallets (like the ones of exchanges or heavy users) when they have a odd distribution of input amount sizes.

**Pull request merged**